feat: add `zone` selection to support a box like selection from given selection

Cloudac7 opened the pull request:
Description:

It introduces a new geometry selection, as described in #4323.
A new selection language is defined in format:
```
box dimension(s) d1_min d1_max (d2_min d2_max) (d3_min d3_max) selection
```
It could pick a zone from the center of geometry from selection.

Changes made in this Pull Request:

 - Add new geometry selection: box


PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

## Developers certificate of origin
- [x] I certify that this contribution is covered by the LGPLv2.1+ license as defined in our [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE) and adheres to the [**Developer Certificate of Origin**](https://developercertificate.org/).


<!-- readthedocs-preview mdanalysis start -->
----
:books: Documentation preview :books:: https://mdanalysis--4324.org.readthedocs.build/en/4324/

<!-- readthedocs-preview mdanalysis end -->

Cloudac7 committed:
feat: add box selection

doc: add doc about box selection

fix: reformat with darker and flake8

pep8speaks commented:
Hello @Cloudac7! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`package/MDAnalysis/core/groups.py`](https://github.com/MDAnalysis/mdanalysis/blob/3f4f6199ce96c2ba7c0dc06974bbf43a139b0501/package/MDAnalysis/core/groups.py):

> [Line 3274:80](https://github.com/MDAnalysis/mdanalysis/blob/3f4f6199ce96c2ba7c0dc06974bbf43a139b0501/package/MDAnalysis/core/groups.py#L3274): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (99 > 79 characters)
> [Line 3275:80](https://github.com/MDAnalysis/mdanalysis/blob/3f4f6199ce96c2ba7c0dc06974bbf43a139b0501/package/MDAnalysis/core/groups.py#L3275): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 3278:75](https://github.com/MDAnalysis/mdanalysis/blob/3f4f6199ce96c2ba7c0dc06974bbf43a139b0501/package/MDAnalysis/core/groups.py#L3278): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 3279:78](https://github.com/MDAnalysis/mdanalysis/blob/3f4f6199ce96c2ba7c0dc06974bbf43a139b0501/package/MDAnalysis/core/groups.py#L3279): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

* In the file [`package/MDAnalysis/core/selection.py`](https://github.com/MDAnalysis/mdanalysis/blob/3f4f6199ce96c2ba7c0dc06974bbf43a139b0501/package/MDAnalysis/core/selection.py):

> [Line 562:80](https://github.com/MDAnalysis/mdanalysis/blob/3f4f6199ce96c2ba7c0dc06974bbf43a139b0501/package/MDAnalysis/core/selection.py#L562): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 567:80](https://github.com/MDAnalysis/mdanalysis/blob/3f4f6199ce96c2ba7c0dc06974bbf43a139b0501/package/MDAnalysis/core/selection.py#L567): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 572:80](https://github.com/MDAnalysis/mdanalysis/blob/3f4f6199ce96c2ba7c0dc06974bbf43a139b0501/package/MDAnalysis/core/selection.py#L572): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)



* In the file [`testsuite/MDAnalysisTests/core/test_atomselections.py`](https://github.com/MDAnalysis/mdanalysis/blob/3f4f6199ce96c2ba7c0dc06974bbf43a139b0501/testsuite/MDAnalysisTests/core/test_atomselections.py):

> [Line 823:80](https://github.com/MDAnalysis/mdanalysis/blob/3f4f6199ce96c2ba7c0dc06974bbf43a139b0501/testsuite/MDAnalysisTests/core/test_atomselections.py#L823): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 836:40](https://github.com/MDAnalysis/mdanalysis/blob/3f4f6199ce96c2ba7c0dc06974bbf43a139b0501/testsuite/MDAnalysisTests/core/test_atomselections.py#L836): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 837:32](https://github.com/MDAnalysis/mdanalysis/blob/3f4f6199ce96c2ba7c0dc06974bbf43a139b0501/testsuite/MDAnalysisTests/core/test_atomselections.py#L837): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 841:40](https://github.com/MDAnalysis/mdanalysis/blob/3f4f6199ce96c2ba7c0dc06974bbf43a139b0501/testsuite/MDAnalysisTests/core/test_atomselections.py#L841): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 842:32](https://github.com/MDAnalysis/mdanalysis/blob/3f4f6199ce96c2ba7c0dc06974bbf43a139b0501/testsuite/MDAnalysisTests/core/test_atomselections.py#L842): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 846:39](https://github.com/MDAnalysis/mdanalysis/blob/3f4f6199ce96c2ba7c0dc06974bbf43a139b0501/testsuite/MDAnalysisTests/core/test_atomselections.py#L846): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 847:32](https://github.com/MDAnalysis/mdanalysis/blob/3f4f6199ce96c2ba7c0dc06974bbf43a139b0501/testsuite/MDAnalysisTests/core/test_atomselections.py#L847): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 851:39](https://github.com/MDAnalysis/mdanalysis/blob/3f4f6199ce96c2ba7c0dc06974bbf43a139b0501/testsuite/MDAnalysisTests/core/test_atomselections.py#L851): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 852:32](https://github.com/MDAnalysis/mdanalysis/blob/3f4f6199ce96c2ba7c0dc06974bbf43a139b0501/testsuite/MDAnalysisTests/core/test_atomselections.py#L852): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace



##### Comment last updated at 2024-07-25 04:03:36 UTC

Cloudac7 committed:
test: add test for orthogonal distance
style: reformat as pep8

Cloudac7 committed:
test: add test in triclinic cell for box selection
test: add error catching for box selection
fix: prevent abnormal input in direction

IAlibay commented on testsuite/MDAnalysisTests/core/test_atomselections.py:
```diff
@@ -1270,16 +1322,39 @@ def test_similarity_selection_icodes(u_pdb_icodes, selection, n_atoms):
-    'byres index 0', 'same resid as index 0',
-])
+
+@pytest.mark.parametrize(
```
We generally ask folks to not reformat large bits of code during their PRs as it makes it very hard to review what actually changed. Could you point out which bits you've added to here?

Cloudac7 replied:
It is an reformat came out directly from execution `darker -r upstream/develop testsuite/MDAnalysisTests -L flake8`, after I added a test script `box xz 3 2 4 -5 index 0` into the empty group test. 
Sorry for not aware of the unexpected modification done by darker.

IAlibay commented on package/doc/sphinx/source/conf.py:
```diff
@@ -14,6 +14,8 @@
 import sys
 import os
 import datetime
+sys.path.insert(0, os.path.abspath('../../..'))
```
Could you explain why this is being added?

Cloudac7 replied:
Sorry for forgetting remove the line. 
I am developing on MacOS platform and thus the compiling of documentation with original code would raise an ModuleNotFoundError: `No module named 'MDAnalysis'`. My practice followed the instruction of latest documentation for developing, and no solution found. After adding the line in more upper position, `sphinx-build` could work.

IAlibay commented on package/doc/sphinx/source/documentation_pages/selections.rst:
```diff
@@ -277,6 +277,23 @@ cyzone *externalRadius* *zMax* *zMin* *selection*
        relative to the COG of *selection*, instead of absolute z-values
        in the box.
 
+box *dimension(s)* *d1_Max* *d1_Min* (*d2_Max* *d2_Min*) (*d3_Max* *d3_Min*) *selection*
```
I'm very likely missing something, but I don't know if I fully understand how this differs from using `prop` - could you maybe explain the benefits of this new selection over using a succession of `prop`?

Cloudac7 replied:
Well, `prop` is used to define a static zone in each axes, in my understanding. The new defined `box` selection do similar thing like which `sphzone`, `cyzone` or `cylayer` do, thus could define a zone according to another selection which could also be dynamical when `updating=True`.

orbeckst replied:
See @Cloudac7 's description in https://github.com/MDAnalysis/mdanalysis/issues/4323#issuecomment-1782184664 , which looks sensible to me. If we have cyzone then we can also have box.

richardjgowers replied:
this would make more sense to me as `min` to `max` values rather than the opposite.  E.g. `box z 0 8 bilayer` to select a slice of water above a bilayer and `box z -8 0` to get the other half.

orbeckst replied:
@richardjgowers I had the same thoughts but unfortunately [cyzone and friends](https://docs.mdanalysis.org/stable/documentation_pages/selections.html#geometric) started this inverse "max min" syntax. I feel consistency across the `*zone` and `*layer` selections is more important here than purity. 

(I mean I am _really_ looking for someone to convince me we should do `min max` here ;-) because I know that I am going to mistype it...)

IAlibay commented on package/CHANGELOG:
```diff
@@ -42,6 +42,7 @@ Enhancements
     (Issue #3994, PR #4281)  
   * Add support for reading chainID info from Autodock PDBQT files (Issue #4207,
     PR #4284)
+  * Add a new geometric selection: box (Issue #4323, PR #4324)
```
We typically go new entries first, could you add this on line 33 instead?

Also please add yourself to the list of authors for 2.7.0 on line 18.

IAlibay reviewed (commented):
Apologies for the slow reviews @Cloudac7

I've added some initial comments.

@MDAnalysis/coredevs could someone please take on the responsibility of reviewing this PR?

Cloudac7 committed:
fix: remove extra line in sphinx conf.py
style: reformat CHANGELOG

richardjgowers commented on package/MDAnalysis/core/selection.py:
```diff
@@ -537,6 +537,101 @@ def _apply(self, group):
+        super().__init__(parser, tokens)
+        self.periodic = parser.periodic
+        self.direction = tokens.popleft()
+        if self.direction not in self.combination:
```
can we set `xmin` `xmax` etc here to `None` then optionally set them to something else in the `else:` branch below?  I think this would simplify the `__getattribute__` usage below

richardjgowers commented on package/MDAnalysis/core/selection.py:
```diff
@@ -537,6 +537,101 @@ def _apply(self, group):
+                # Orthogonal version
+                vecs -= box[:3] * np.rint(vecs / box[:3])
+            else:
+                # Triclinic version
+                tribox = group.universe.trajectory.ts.triclinic_dimensions
+                vecs -= tribox[2] * np.rint(vecs[:, 2] / tribox[2][2])[:, None]
+                vecs -= tribox[1] * np.rint(vecs[:, 1] / tribox[1][1])[:, None]
+                vecs -= tribox[0] * np.rint(vecs[:, 0] / tribox[0][0])[:, None]
```
this isn't technically correct for some triclinic boxes with some shifts.

richardjgowers replied:
I think you'll want this function: https://github.com/MDAnalysis/mdanalysis/blob/c36313e4e849bc37033a142454e9e593db2b3272/package/MDAnalysis/lib/distances.py#L1791

But it won't handle some, but not all, box dimensions being given (i.e. semi periodic calculation)

Cloudac7 committed:
fix: use `minimize_vectors` to calculate min vec

Cloudac7 committed:
fix: set default dmin and dmax for box selection

Cloudac7 committed:
doc: format docstring as well as doc for selections

RMeli commented on package/MDAnalysis/core/groups.py:
```diff
@@ -3111,6 +3111,26 @@ def select_atoms(self, sel, *othersel, periodic=True, rtol=1e-05,
+                *dimensions* Specifies which dimension(s) to apply
+                the box selection on. Can be ``x``, ``y``, ``z``,
+                or any combination like ``xy``, ``yz``, ``zx``, ``xyz``
+                (up to 3 characters). *dN_min*, *dN_max* are the minimum and
+                maximum bounds along the first specified dimension.
```
```suggestion
                (up to 3 characters). *dN_min*, *dN_max* are the minimum and
                maximum bounds along each specified dimension.
```
?

RMeli commented on package/MDAnalysis/core/groups.py:
```diff
@@ -3111,6 +3111,26 @@ def select_atoms(self, sel, *othersel, periodic=True, rtol=1e-05,
                 radius 5, external radius 10 centered on the COG. In z, the
                 cylinder extends from 10 above the COG to 8 below. Positive
                 values for *zMin*, or negative ones for *zMax*, are allowed.
+            box *dimensions* *dN_min* *dN_max* *selection*
```
```suggestion
            box *dimensions* *dN_min* *dN_max* [*dN_min* *dN_max*] [*dN_min* *dN_max*] *selection*
```
Specify also the optional arguments (depending on `dimension`)?

RMeli commented on package/MDAnalysis/core/selection.py:
```diff
@@ -537,6 +537,103 @@ def _apply(self, group):
+class BoxSelection(Selection):
+    token = "box"
+    precedence = 1
+    combination = [
+        "x",
+        "y",
+        "z",
+        "xy",
+        "xz",
+        "yz",
+        "yx",
+        "zx",
+        "zy",
+        "xyz",
+        "xzy",
+        "yxz",
+        "yzx",
+        "zxy",
+        "zyx",
+    ]
```
```suggestion
    combination = [''.join(s) for i in range(1, 4) for s in itertools.permutations("xyz", i)]
```
Just for fun. Probably less readable than what you wrote. 

But the question is, do we really need both `xy` and `yx`? What's the use case of having all permutations as opposed to only combinations?

```suggestion
    combination = [''.join(s) for i in range(1, 4) for s in itertools.combinations("xyz", i)]
```

Nitpick: the variable is called "combination", so the name might have to change to "permutation" ("permutations"?) if that's the actual use case.

Cloudac7 replied:
The extensive list of all permutations of `x`, `y`, and `z` is designed to avoid the need to calculate the O(3!) = O(1) permutations each time a selection is executed. This has been reformatted by `darker` to comply with PEP8 standards. It's a minor optimization that can be overlooked, though. 🤣

Regarding the latter point, my original intention was to enhance the flexibility of the selection by making both `yx` and `xy` acceptable. However, this inadvertently added extra complexity. You're correct in suggesting that it would be better to have users write in the sequence of `x -> y -> z`. This approach would eliminate the need for additional processing.

RMeli commented on package/MDAnalysis/core/selection.py:
```diff
@@ -537,6 +537,103 @@ def _apply(self, group):
+        self.zmin, self.zmax = None, None
+        if self.direction not in self.combination:
+            raise ValueError(
+                "The direction '{}' is not valid. Must be combination of {}"
+                "".format(self.direction, ["x", "y", "z"])
```
```suggestion
                f"The direction '{self.direction}' is not valid. Must be combination of {axis_map}"
```

Cloudac7 committed:
Squashed commit of the following:

commit 8d7f2ce009499430e97dacd6a4e765c2dc6cbd56
Author: Futaki Haduki <812556867@qq.com>
Date:   Thu Nov 9 17:23:57 2023 +0800

    Update package/MDAnalysis/core/groups.py

    Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

commit cfdec5b44f911216ed30fae32fe472debe790afe
Author: Futaki Haduki <812556867@qq.com>
Date:   Thu Nov 9 17:23:44 2023 +0800

    Update package/MDAnalysis/core/groups.py

    Co-authored-by: Rocco Meli <r.meli@bluemail.ch>

Cloudac7 committed:
fix: rollback unexpected change in unittest

orbeckst commented:
@richardjgowers I am tentatively assigning the PR to you for shepherding it through. If it's too much then please unassign yourself and ping me. Thanks.

orbeckst committed:
Merge branch 'develop' into box-selection

orbeckst commented:
@richardjgowers could you have a look if @Cloudac7 addressed the reviewers' comments? Thanks!

orbeckst commented:
@richardjgowers could you please have another look?

today on discord someone asked for this feature so I was reminded of this PR…

orbeckst commented:
@Cloudac7 could you please resolve the conflicts?

Sorry that this has been very slow from our end.

Cloudac7 committed:
Merge branch 'develop' into box-selection

richardjgowers commented on package/MDAnalysis/core/selection.py:
```diff
@@ -537,6 +537,96 @@ def _apply(self, group):
+        }
+
+        if self.periodic and group.dimensions is not None:
+            box = group.dimensions[:3]
```
I'm not sure this is correct for triclinic boxes, I think instead you want the trace of the 3x3 vector representation

Cloudac7 replied:
In fact, I'm not sure if the check necessary. In my opinion, it might be nature to select a zone a little larger than the unit cell, yielding all of the atoms in the unit cell on the corresponding direction. Although it would be misleading because of PBC.

richardjgowers commented on package/MDAnalysis/core/groups.py:
```diff
@@ -3271,6 +3271,25 @@ def select_atoms(self, sel, *othersel, periodic=True, rtol=1e-05,
                 radius 5, external radius 10 centered on the COG. In z, the
                 cylinder extends from 10 above the COG to 8 below. Positive
                 values for *zMin*, or negative ones for *zMax*, are allowed.
+            box *dimensions* *dN_min* *dN_max* [*dN_min* *dN_max*] [*dN_min* *dN_max*] *selection*
```
We already have sphzone so maybe this could be called zone?

richardjgowers reviewed (commented):
Looks good, couple tweaks to go

Cloudac7 committed:
test: fix wrong error message

Cloudac7 committed:
refactor: rename the selection as `zone`;
feat: remove check on unit cell for `zone`

Cloudac7 committed:
test: fix typo

Cloudac7 committed:
test: add tests of limits for `zone`
